Web Inspector: Heap: custom elements should each have their own category instead of `HTMLElement`

dcrousso opened the pull request:
#### 5c1282ed9528b7276478130fbaa4d369dea0d5cc
<pre>
Web Inspector: Heap: custom elements should each have their own category instead of `HTMLElement`
<a href="https://bugs.webkit.org/show_bug.cgi?id=289935">https://bugs.webkit.org/show_bug.cgi?id=289935</a>

Reviewed by Ryosuke Niwa.

Custom JS `class` instances are given a separate top-level category in the &quot;Instances&quot; view, so custom elements should too (instead of being lumped into `HTMLElement`) since they&apos;re pretty much the same concept.

* Source/JavaScriptCore/heap/HeapSnapshotBuilder.h:
(JSC::HeapSnapshotBuilder::Client::heapSnapshotBuilderIgnoreNode): Added.
(JSC::HeapSnapshotBuilder::Client::heapSnapshotBuilderOverrideClassName): Added.
(JSC::HeapSnapshotBuilder::Client::heapSnapshotBuilderIsElement): Added.
* Source/JavaScriptCore/heap/HeapSnapshotBuilder.cpp:
(JSC::HeapSnapshotBuilder::descriptionForNode): Renamed from `descriptionForCell`.
(JSC::HeapSnapshotBuilder::json):
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.cpp:
(Inspector::InspectorHeapAgent::snapshot):
(Inspector::InspectorHeapAgent::heapSnapshotBuilderIgnoreNode): Added.
* Source/WebCore/inspector/agents/page/PageHeapAgent.h:
* Source/WebCore/inspector/agents/page/PageHeapAgent.cpp:
(WebCore::PageHeapAgent::heapSnapshotBuilderOverrideClassName): Added.
(WebCore::PageHeapAgent::heapSnapshotBuilderIsElement): Added.
Add a way for callers to override the what the class name is for a given `HeapSnapshotNode`.
Additionally, since DOM nodes are treated differently (i.e. they&apos;re rendered as HTML instead of a JS object), add a way to `HeapSnapshotNode` as such (i.e. `NodeFlags::ElementSubtype`).

* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
(HeapSnapshot):
(HeapSnapshot.updateCategoriesAndMetadata):
(HeapSnapshot.prototype.serializeNode):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotNodeProxy.js:
(WI.HeapSnapshotNodeProxy):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotClassDataGridNode.js:
(WI.HeapSnapshotClassDataGridNode.prototype.createCellContent):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotClusterContentView.js:
(WI.HeapSnapshotClusterContentView.iconStyleClassNameForClassName):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotDataGridTree.js:
(WI.HeapSnapshotInstancesDataGridTree.prototype.populateTopLevel):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.prototype.createCellContent):
(WI.HeapSnapshotInstanceDataGridNode.prototype._getRemoteObject):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateWindowPreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populatePreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler.appendPathRow):
Use an element icon instead of an object icon if the `HeapSnapshotNode` is marked as a DOM node (i.e. `NodeFlags::ElementSubtype`).

* Source/WebInspectorUI/UserInterface/Views/FormattedValue.js:
(WI.FormattedValue.createObjectPreviewOrFormattedValueForObjectPreview):
(WI.FormattedValue.createObjectPreviewOrFormattedValueForRemoteObject):
* Source/WebInspectorUI/UserInterface/Views/ObjectPreviewView.js:
(WI.ObjectPreviewView):
* Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
(WI.ConsoleMessageView.prototype._appendFormattedArguments):
Pass along an `options = {}` so that a `remoteObjectAccessor` can be defined for highlighting DOM nodes (i.e. `NodeFlags::ElementSubtype`) in the heap snapshot.

* LayoutTests/inspector/heap/snapshot.html:
* LayoutTests/inspector/heap/snapshot-expected.txt:

Canonical link: <a href="https://commits.webkit.org/292456@main">https://commits.webkit.org/292456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c529ee5f0c6593550dbb675111dd23e886d2a93

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96123 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15737 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101188 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46632 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98168 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24170 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/73277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/30492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99126 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/12009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/86825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53612 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11759 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4581 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45967 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/88796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/4683 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103213 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94744 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23190 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23441 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81686 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20490 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26301 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23153 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/118221 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22812 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26292 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24553 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/JavaScriptCore/heap/HeapSnapshotBuilder.h:
```diff
@@ -131,26 +131,38 @@ class JS_EXPORT_PRIVATE HeapSnapshotBuilder final : public HeapAnalyzer {
-    
+
     HeapProfiler& m_profiler;
+    Client* m_client { nullptr };
```
Please use `WeakPtr` or `CheckedPtr` instead.

rniwa commented on Source/WebCore/inspector/agents/page/PageHeapAgent.cpp:
```diff
@@ -58,6 +62,35 @@ Inspector::Protocol::ErrorStringOr<void> PageHeapAgent::disable()
+            if (element->isDefinedCustomElement()) {
+                if (RefPtr customElementRegistry = element->customElementRegistry()) {
+                    if (RefPtr customElementInterface = customElementRegistry->findInterface(element)) {
+                        if (JSC::JSObject* customElementConstructorObject = customElementInterface->constructor()) {
```
This is a roundabout way to get to the constructor given we already have `JSCell*` but okay.

dcrousso replied:
i got this by working backwards from what was provided to `CustomElementRegistry.prototype.define`.  happy to do something else if you think there's a more efficient approach :)

rniwa replied:
Since we already have `JSCell`, we can just get the prototype then its constructor but this code is fine as is.

rniwa commented on Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
```diff
@@ -102,7 +104,7 @@ HeapSnapshot = class HeapSnapshot
 
         let {version, type, nodes, nodeClassNames, edges, edgeTypes, edgeNames, roots, labels} = json;
-        console.assert(version === 1 || version === 2, "Expect JavaScriptCore Heap Snapshot version 1 or 2");
+        console.assert(version === 1 || version === 2 || version === 3, version);
```
It seems like the this assertion message got worse?
How about this?
```
`Expect JavaScriptCore Heap Snapshot version 1 - 3 but got ${version}`
```

dcrousso replied:
typically the `console.assert(...)` in Web Inspector use only the tested values as the message as engineers will only see these in inspector^2, meaning there'd be a link to the source code location where they can figure out what was intended by looking at the condition (i.e. the reason for providing the tested values is to know how exactly the assertion failed)

rniwa replied:
ok.

dcrousso committed:
Web Inspector: Heap: custom elements should each have their own category instead of `HTMLElement`
https://bugs.webkit.org/show_bug.cgi?id=289935

Reviewed by Ryosuke Niwa.

Custom JS `class` instances are given a separate top-level category in the "Instances" view, so custom elements should too (instead of being lumped into `HTMLElement`) since they're pretty much the same concept.

* Source/JavaScriptCore/heap/HeapSnapshotBuilder.h:
(JSC::HeapSnapshotBuilder::Client::heapSnapshotBuilderIgnoreNode): Added.
(JSC::HeapSnapshotBuilder::Client::heapSnapshotBuilderOverrideClassName): Added.
(JSC::HeapSnapshotBuilder::Client::heapSnapshotBuilderIsElement): Added.
* Source/JavaScriptCore/heap/HeapSnapshotBuilder.cpp:
(JSC::HeapSnapshotBuilder::descriptionForNode): Renamed from `descriptionForCell`.
(JSC::HeapSnapshotBuilder::json):
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.cpp:
(Inspector::InspectorHeapAgent::snapshot):
(Inspector::InspectorHeapAgent::heapSnapshotBuilderIgnoreNode): Added.
* Source/WebCore/inspector/agents/page/PageHeapAgent.h:
* Source/WebCore/inspector/agents/page/PageHeapAgent.cpp:
(WebCore::PageHeapAgent::heapSnapshotBuilderOverrideClassName): Added.
(WebCore::PageHeapAgent::heapSnapshotBuilderIsElement): Added.
Add a way for callers to override the what the class name is for a given `HeapSnapshotNode`.
Additionally, since DOM nodes are treated differently (i.e. they're rendered as HTML instead of a JS object), add a way to `HeapSnapshotNode` as such (i.e. `NodeFlags::ElementSubtype`).

* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
(HeapSnapshot):
(HeapSnapshot.updateCategoriesAndMetadata):
(HeapSnapshot.prototype.serializeNode):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotNodeProxy.js:
(WI.HeapSnapshotNodeProxy):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotClassDataGridNode.js:
(WI.HeapSnapshotClassDataGridNode.prototype.createCellContent):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotClusterContentView.js:
(WI.HeapSnapshotClusterContentView.iconStyleClassNameForClassName):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotDataGridTree.js:
(WI.HeapSnapshotInstancesDataGridTree.prototype.populateTopLevel):
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.prototype.createCellContent):
(WI.HeapSnapshotInstanceDataGridNode.prototype._getRemoteObject):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateWindowPreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populatePreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler.appendPathRow):
Use an element icon instead of an object icon if the `HeapSnapshotNode` is marked as a DOM node (i.e. `NodeFlags::ElementSubtype`).

* Source/WebInspectorUI/UserInterface/Views/FormattedValue.js:
(WI.FormattedValue.createObjectPreviewOrFormattedValueForObjectPreview):
(WI.FormattedValue.createObjectPreviewOrFormattedValueForRemoteObject):
* Source/WebInspectorUI/UserInterface/Views/ObjectPreviewView.js:
(WI.ObjectPreviewView):
* Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
(WI.ConsoleMessageView.prototype._appendFormattedArguments):
Pass along an `options = {}` so that a `remoteObjectAccessor` can be defined for highlighting DOM nodes (i.e. `NodeFlags::ElementSubtype`) in the heap snapshot.

* LayoutTests/inspector/heap/snapshot.html:
* LayoutTests/inspector/heap/snapshot-expected.txt:

Canonical link: https://commits.webkit.org/292456@main
